src/binary/load_kernel.rs

```diff
@@ -220,47 +227,16 @@ where
             // the remaining part of the frame since the frame is no longer shared with other
             // segments now.
 
-            // calculate the frame where the last segment page is mapped
-            let orig_frame: PhysFrame =
-                PhysFrame::containing_address(phys_start_addr + file_size - 1u64);
-            // allocate a new frame to replace `orig_frame`
-            let new_frame = self.frame_allocator.allocate_frame().unwrap();
-
-            // zero new frame, utilizing that it's identity-mapped
-            {
-                let new_frame_ptr = new_frame.start_address().as_u64() as *mut PageArray;
-                unsafe { new_frame_ptr.write(ZERO_ARRAY) };
-            }
-
-            // copy the data bytes from orig_frame to new_frame
-            {
-                log::info!("Copy contents");
-                let orig_bytes_ptr = orig_frame.start_address().as_u64() as *mut u8;
-                let new_bytes_ptr = new_frame.start_address().as_u64() as *mut u8;
-
-                for offset in 0..(data_bytes_before_zero as isize) {
-                    unsafe {
-                        let orig_byte = orig_bytes_ptr.offset(offset).read();
-                        new_bytes_ptr.offset(offset).write(orig_byte);
-                    }
-                }
-            }
-
-            // remap last page from orig_frame to `new_frame`
-            log::info!("Remap last page");
             let last_page = Page::containing_address(virt_start_addr + file_size - 1u64);
-            self.page_table
-                .unmap(last_page.clone())
-                .map_err(|_err| "Failed to unmap last segment page because of bss memory")?
-                .1
-                .ignore();
-            let flusher = unsafe {
-                self.page_table
-                    .map_to(last_page, new_frame, segment_flags, self.frame_allocator)
+            let new_frame = unsafe { self.make_mut(last_page) };
+            let new_bytes_ptr = new_frame.start_address().as_u64() as *mut u8;
+            unsafe {
+                core::ptr::write_bytes(
+                    new_bytes_ptr,
+                    0,
+                    (Size4KiB::SIZE - data_bytes_before_zero) as usize,
+                );
             }
```

Comment:
Shouldn't we zero the _last_ bytes of the frame here?

Reply:
Yes, good catch!